include/range/v3/index.hpp

```diff
@@ -0,0 +1,95 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Eric Niebler 2014
+//  Copyright Gonzalo Brito Gadeschi 2017
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGES_V3_INDEX_HPP
+#define RANGES_V3_INDEX_HPP
+
+#include <range/v3/range_fwd.hpp>
+#include <range/v3/begin_end.hpp>
+#include <range/v3/range_traits.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/distance.hpp>
+#include <range/v3/utility/static_const.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        /// \cond
+        namespace index_detail
+        {
+            template<typename Rng, typename T>
+            using Concept = meta::and_<
+                RandomAccessRange<Rng>,
+                // Only evaluate this one if the previous one succeeded
+                meta::lazy::invoke<
+                    meta::compose<
+                        meta::bind_front<meta::quote<ConvertibleTo>, T>,
+                        meta::quote<range_difference_type_t>>,
+                    Rng>>;
+        }  // namespace index_detail
+        /// \endcond
+
+        /// Unchecked indexed range access.
+        ///
+        /// \ingroup group-core
+        struct index_fn
+        {
+            /// \return `begin(rng)[n]`
+            template<typename Rng,
+                CONCEPT_REQUIRES_(RandomAccessRange<Rng>())>
+            RANGES_CXX14_CONSTEXPR
+            auto operator()(Rng &&rng, range_difference_type_t<Rng> n) const
```

Comment:
RANGES_EXPECT failed in some builds, but i've added another `noexcept(...)` clause.